Refactor QueryManager to make better use of observables and enforce fetchPolicy more reliably

benjamn opened the pull request:
Fixes #5991, #5790, and multiple other related issues (full list TBD).

This is a large pull request, representing about a month and a half of work, but the central idea is relatively simple: the most important job of the `QueryManager` is to tirelessly enforce the requested `FetchPolicy` (`cache-first`, `cache-and-network`, `network-only`, etc.) for each query that it manages. Previously, the `QueryManager` sometimes failed to fulfill this mission, because the logic for interpreting `FetchPolicy` values was scattered across the codebase in an ad hoc manner, rather than centralized in one place.

For example, the default `cache-first` fetch policy should reliably trigger a network request if the relevant cache data disappears or becomes incomplete (#5991), but the code that detected that incompleteness (`queryListenerForObserver`) was not responsible for triggering network requests, and giving it the ability to call `fetchQuery` would have required yet another one-off, decentralized interpretation of `options.fetchPolicy`.

The scattering of `FetchPolicy` logic made it very difficult for anyone (including me) to understand how each fetch policy works at a glance, adding risk to any changes that involved fetch policies.

Thanks to the changes in this PR, the logic for interpreting fetch policies has been centralized into a single `switch` statement, which I think reads almost as clearly as pseudocode:

```ts
switch (fetchPolicy) {
  default: case "cache-first": {
    const diff = readCache();

    if (diff.complete) {
      return [
        resultsFromCache(diff, queryInfo.markReady()),
      ];
    }

    if (returnPartialData) {
      return [
        resultsFromCache(diff),
        resultsFromLink(true),
      ];
    }

    return [
      resultsFromLink(true),
    ];
  }

  case "cache-and-network": {
    const diff = readCache();

    if (diff.complete || returnPartialData) {
      return [
        resultsFromCache(diff),
        resultsFromLink(true),
      ];
    }

    return [
      resultsFromLink(true),
    ];
  }

  case "cache-only":
    return [
      resultsFromCache(readCache(), queryInfo.markReady()),
    ];

  case "network-only":
    return [resultsFromLink(true)];

  case "no-cache":
    return [resultsFromLink(false)];

  case "standby":
    return [];
}
```

The arrays returned by this code contain zero, one, or two observables, derived from the cache and/or the network link. The observables are concatenated together to form a single observable (specifically, a `Concast` observable) whose results are then consumed by the `ObservableQuery` (with any adjacent duplicate results skipped, as usual). At any point in time, most often when cache data has changed, or the application specifically requested a refetch, the `ObservableQuery` can `reobserve` the cache and/or the network by invoking this same code again. In other words, this `switch` statement is now the beating heart of the client codebase, and no data flows through the client without honoring the requested `FetchPolicy`.

This new system (`fetchQueryObservable`) is much more `Observable`-oriented than the previous `fetchQuery` method was, and I see this shift as a huge improvement. One of the reasons the `QueryManager` was so complicated was that the old `fetchQuery` method could return only a single `Promise<ApolloQueryResult<T>>`, even though there was often more than one result to deliver, necessitating a side-delivery mechanism (`broadcastQueries`). By allowing `fetchQueryObservable` to return multiple results using an `Observable`, we no longer have to worry about delivering different results using different mechanisms, since all results flow through a straightforward `Observable` pipeline.

Since this is such a deep refactoring, I found it impossible to keep existing tests passing during the initial exploration. Once I was happy with the shape of the new system, I resumed running and debugging the tests, and rebased any necessary code changes back into my previous commits. There was never any hope of leaving all existing tests unchanged, but I think/hope the test changes I made (see commits towards the end of this PR) are justifiable. The whole point of this refactoring was to make Apollo Client do a better job of fetching data, and "better" sometimes means "different." In any case, if you see something I changed in the tests that worries you, please leave comments/questions in-line.

The most substantial commit is `Replace fetch{Query,Request} with new fetchQueryObservable method` (1ac5d8de4bfd7a17c8f4b65dd58ad43a1e6fdd38). I regret that so much ended up in that commit, but it's also a commit whose parts don't work until you put them all together, so I'm not going to waste time trying to split it up at this point. For what it's worth, [everything that comes before that commit](https://github.com/apollographql/apollo-client/compare/master...f35a3d903142da14d3d75c1f6a458471100cf61d) is "safe" in the sense of keeping all existing tests passing.

I will leave some comments below to draw attention to any interesting/controversial changes.

benjamn commented:
Ugh, GitHub is showing these commits in chronological order, rather than parent-child order. I will reset all the timestamps and force-push momentarily.

benjamn commented:
To save you a click:
```
Test Suites: 3 failed, 56 passed, 59 total
Tests:       6 failed, 6 skipped, 919 passed, 931 total
Snapshots:   36 passed, 36 total
```
I hope it goes without saying that all tests have been passing locally. 🧐

3nvi commented:
Tried npm-installing it through the PR branch and building it locally, but had some issues. Is there a cache we can get a preview version on NPM to test it (whenever it's ready/stable)?

benjamn commented:
@3nvi I've had some luck with running `npm pack` in the `apollo-client/dist` directory after a successful `npm run build`, and then doing
```
npm install ../path/to/apollo-client/dist/apollo

benjamn commented:
Ok, I can reproduce some of the tests failures locally if I use Node 10, which is what our CircleCI job uses. Locally I'd been using Node 12.16.1. No theory yet on what the operative difference between the Node versions might be.

benjamn commented on src/react/data/QueryData.ts:
```diff
@@ -102,9 +102,7 @@ export class QueryData<TData, TVariables> extends OperationData {
-      setTimeout(() => {
-        this.currentObservable?.resetQueryStoreErrors();
-      });
+      this.currentObservable?.resetQueryStoreErrors();
```
@hwillson Can you recall why it seemed like we needed to use `setTimeout` here?

hwillson replied:
@benjamn It was used to make sure errors weren't reset until after they were displayed during render (if there were any). https://github.com/apollographql/apollo-client/pull/6107 removes the need for it though by handling the reset differently.

benjamn replied:
Merged #6107 and rebased this PR against `master`—thanks!

darkbasic commented:
@benjamn I suggest using multiple node versions on CI, one per major release.

benjamn commented on src/core/__tests__/fetchPolicies.ts:
```diff
@@ -382,7 +382,7 @@ describe('cache-and-network', function() {
       } else if (count === 2) {
         expect(result).toEqual({
-          data: dataWithId(1),
+          data: {},
```
This empty object is the incomplete (that is, not even partially complete) cache result that we are delivering only because of the `notifyOnNetworkStatusChange: true` option above. Perhaps we should not even include the `data` property, since the `{}` is pretty useless, and the important information is `loading: true` and `networkStatus: NetworkStatus.setVariables`. Before the changes in this PR, we would often redeliver the previous `data` (in this case, `dataWithId(1)`), which could conceivably be useful for code that uses `data` without checking the `loading` status, but redelivering old data feels less appropriate (to me) in this new `fetchQueryObservable`-based system.

benjamn replied:
That said, `data` here is not necessarily always `{}`, when relevant data can be found in the cache for the given query. So `{}` is consistent with the behavior of `returnPartialData: true`, except for being triggered by `notifyOnNetworkStatusChange` rather than `returnPartialData`.

benjamn commented on src/core/__tests__/scheduler.ts:
```diff
@@ -1,512 +0,0 @@
-  );
-}
-
-describe('QueryScheduler', () => {
```
This test file was excessively (IMO) dependent on private properties of the `QueryManager` that no longer exist, so I decided simply to delete the tests. Polling is now confined to the `Reobserver` class, which should make polling behavior more consistent/reliable and easier to understand. I'm open to reviving these tests if they seem valuable, but they would have to change pretty drastically.

benjamn commented on src/core/types.ts:
```diff
@@ -22,12 +22,6 @@ export type ApolloQueryResult<T> = {
   networkStatus: NetworkStatus;
 };
 
-export enum FetchType {
-  normal = 1,
-  refetch = 2,
-  poll = 3,
```
This `FetchType` enum was only used internally, and seemed duplicative with `NetworkStatus`, so I removed it.

benjamn commented on src/utilities/observables/Concast.ts:
```diff
@@ -0,0 +1,245 @@
+    this.handlers.complete!();
+  }
+
+  // Generic implementations of Observable.prototype methods like map and
+  // filter need to know how to create a new Observable from a Concast.
+  // Those methods assume (perhaps unwisely?) that they can call the
+  // subtype's constructor with an observer registration function, but the
+  // Concast constructor uses a different signature. Defining this
+  // Symbol.species getter function on the Concast constructor function is
+  // a hint to generic Observable code to use the default constructor
+  // instead of trying to do `new Concast(observer => ...)`.
+  static get [Symbol.species]() {
+    return Observable;
+  }
```
I'm a little uneasy about using `Symbol.species` and `Symbol.iterator` in this class, though they both serve important purposes, because they create an additional polyfilling burden for older browsers.

darkbasic replied:
Deciding to support IE11 means having to deal with it until at least October 2025 (2028 if you count Windows Enterprise). Is it really worth it? Even Ionic decided to drop it because the technical gap between IE11 and other browsers was too big. Lack of CSS variables alone is big enough to make developers experience way worse.

benjamn replied:
Resolved for now by 6ff3a5923c94649dee0990150f242ed6d66e6a0b and cfa385d5aa43ef1c5d0c0c20c33b3c6dfc7daa0d, though I genuinely hope @darkbasic is right and I'm just being paranoid.

benjamn commented on src/utilities/observables/Concast.ts:
```diff
@@ -0,0 +1,245 @@
+  // easy way to observe the final state of the Concast.
+  private resolve: (result?: T) => void;
+  private reject: (reason: any) => void;
+  public readonly promise = new Promise<T>((resolve, reject) => {
```
One of my favorite things about `Concast` observables is that each one automatically has a `concast.promise` property, which can be used by code that only cares about the final result, or ignored by code that uses the `Observable` interface.

hwillson replied:
This is very cool!!

benjamn commented on src/core/QueryManager.ts:
```diff
@@ -506,82 +376,6 @@ export class QueryManager<TStore> {
-
-        observer.next && observer.next(result);
-
-      } else if (process.env.NODE_ENV !== 'production' &&
-                 isNonEmptyArray(diff.missing) &&
-                 !equal(diff.result, {})) {
-        invariant.warn(`Missing cache result fields: ${
-          diff.missing.map(m => m.path.join('.')).join(', ')
-        }`, diff.missing);
```
Note/`TODO`: We have temporarily lost these warnings (introduced by #6058). I need to rethink how we report these errors in the new pipeline. One possibility: maybe we just add an `errors` property to `ApolloQueryResult<TData>`, which includes any missing field errors, so that the errors are inspectable but not quite so noisy in the console?

benjamn replied:
Rather than adding more properties to `ApolloQueryResult`, I decided just to restore the previous warning functionality: 9e563867768603007f465b4d4197cfdd47ad7196

benjamn commented on src/core/QueryManager.ts:
```diff
@@ -911,74 +687,15 @@ export class QueryManager<TStore> {
     this.queries.delete(queryId);
   }
 
-  public getCurrentQueryResult<T>(
```
This `getCurrentQueryResult` method (along with `getQueryWithPreviousResult`) moved into the `ObservableQuery` class, but I'm not thrilled with either of them, and I hope we can find a way to eliminate them in the future.

benjamn commented on src/core/QueryManager.ts:
```diff
@@ -1025,31 +742,25 @@ export class QueryManager<TStore> {
-            next: cleanup,
-            error: cleanup,
-            complete: cleanup,
+          cc.cleanup(() => {
```
The `Concast#cleanup` function is useful because it does not require subscribing to the `Concast`, which can be risky when the cleanup subscription ends before other subscriptions have started, which tends to prematurely terminate the observable.

benjamn commented on src/core/QueryManager.ts:
```diff
@@ -1068,114 +779,268 @@ export class QueryManager<TStore> {
+  private getResultsFromLink<TData, TVars>(
+    queryInfo: QueryInfo,
+    allowCacheWrite: boolean,
+    options: Pick<WatchQueryOptions<TVars>,
+      | "variables"
+      | "context"
+      | "fetchPolicy"
+      | "errorPolicy">,
```
I've become fond of using the `Pick` generic type to limit the required fields of object parameters using the type system, while continuing to pass in the whole `WatchQueryOptions` object.

benjamn commented on src/core/QueryManager.ts:
```diff
@@ -1068,114 +779,268 @@ export class QueryManager<TStore> {
+        typeof oldNetworkStatus === "number" &&
+        oldNetworkStatus !== networkStatus &&
+        isNetworkRequestInFlight(networkStatus)) {
+      if (fetchPolicy !== "cache-first") {
+        fetchPolicy = "cache-and-network";
+      }
+      returnPartialData = true;
```
In order to force delivery of an incomplete cache result with `loading: true`, we tweak the `fetchPolicy` to include the cache, and pretend that `returnPartialData` was enabled.

benjamn commented on src/core/Reobserver.ts:
```diff
@@ -0,0 +1,142 @@
+
+    const concast = this.fetch(this.options, newNetworkStatus);
+
+    if (this.concast) {
+      // We use the {add,remove}Observer methods directly to avoid
+      // wrapping observer with an unnecessary SubscriptionObserver
+      // object, in part so that we can remove it here without triggering
+      // any unsubscriptions, because we just want to ignore the old
+      // observable, not prematurely shut it down, since other consumers
+      // may be awaiting this.concast.promise.
+      this.concast.removeObserver(this.observer, true);
+    }
+
+    concast.addObserver(this.observer);
```
This code is important because it ensures `this.observer` is only ever subscribed to (at most) one `Concast`, and never misses any messages, because the `removeObserver` call comes immediately before the `addObserver` call.

danReynolds commented on src/core/Reobserver.ts:
```diff
@@ -0,0 +1,142 @@
+    if (newOptions) {
+      this.updateOptions(newOptions);
+    } else {
+      // When we call this.updatePolling(newOptions) in the branch above,
```
```suggestion
      // When we call this. updateOptions(newOptions) in the branch above,
```

mogelbrod commented:
Any chance this PR might address the following issue? #5239 
Potential fix available here: https://github.com/apollographql/apollo-client/pull/6023

benjamn commented:
@jcreighton @hwillson In case you were wondering/worried about merge conflicts from #6115, I just rebased this PR, and it was all pretty straightforward. It helped that I could just default to taking my own changes, and then re-fix any remaining type errors. Rebasing definitely would have been harder in the other direction (if this PR had been merged first).

benjamn commented on src/core/QueryInfo.ts:
```diff
@@ -112,39 +106,23 @@ export class QueryInfo {
 
-    return this.diff;
-  }
+    (this as any).observableQuery = oq;
```
If I'm going to make something `public`, I prefer it to be outwardly `readonly` if possible, which requires internal overrides like this. I wish TypeScript had an annotation that reflected public immutability with private mutability, without having to use a getter method.

hwillson replied:
Sounds good! (and that would be an awesome add to TS for sure)

benjamn commented on src/core/QueryManager.ts:
```diff
@@ -1068,114 +779,268 @@ export class QueryManager<TStore> {
+    return concast;
+  }
+
+  private fetchQueryHelper<TData, TVars>(
```
How about `fetchQueryByPolicy`?

hwillson replied:
That's perfect @benjamn!

benjamn commented on src/core/ObservableQuery.ts:
```diff
@@ -293,36 +296,34 @@ export class ObservableQuery<
+      // instead of network-only), because we allow the caller of
+      // fetchMore to provide an updateQuery callback that determines how
+      // the data gets written to the cache.
+      fetchPolicy: "no-cache",
```
We do still call `this.updateQuery`, even if no special `updateQuery` function was provided, so the data get written to the cache, despite this `no-cache` policy. Do you think we should avoid calling `this.updateQuery` (and thus prevent writing to the cache) if the given `fetchPolicy` (that is, the one we're overriding here) was `no-cache`?

Generally speaking, I think we should pay special attention to `fetchMore` before the AC3 launch, as it's been the source of a number of [issues](https://github.com/apollographql/apollo-client/issues?q=is%3Aissue+is%3Aopen+fetchMore), both recently and in the past.

benjamn replied:
We do still call `this.updateQuery`, even if no special `updateQuery` function was provided, so the data get written to the cache, despite this `no-cache` policy. Do you think we should avoid calling `this.updateQuery` (and thus prevent writing to the cache) if the given `fetchPolicy` (that is, the one we're overriding here) was `no-cache`?

Generally speaking, I think we should pay special attention to `fetchMore` before the AC3 launch, as it's been the source of a number of [issues](https://github.com/apollographql/apollo-client/issues?q=is%3Aissue+is%3Aopen+fetchMore), both recently and in the past.

hwillson replied:
Ahh got it, thanks @benjamn. Hmm - great question. I think writing to the cache after a `fetchMore` call (even if an `updateQuery` isn't provided) is what people are used to. I was just thrown off by seeing the "no-cache" here. I think we're okay to keep this as you have it, but I definitely agree we should pay special attention to `fetchMore` issues (I'll dig through the issue backlog as well, to look at some of the common trip ups we can help address).

benjamn commented on src/core/ObservableQuery.ts:
```diff
@@ -457,52 +425,81 @@ export class ObservableQuery<
-      this.options,
-    ) as Promise<ApolloQueryResult<TData>>;
+    let { fetchPolicy } = this.options;
+    if (fetchPolicy !== 'cache-first' &&
+        fetchPolicy !== 'no-cache' &&
+        fetchPolicy !== 'network-only') {
```
I think we need to stick to a `union` of string literals, since the `fetchPolicy` is something that developers provide in the `WatchQueryOptions`.

Honestly, I don't love the replacement of `NetworkStatus` values with numbers, since it requires either memorizing or looking up the numbers when you're debugging, but that's also part of the public API now, so I don't think we should change it.

hwillson replied:
Great point and sounds good to me. And oh yes, I'm constantly cracking open `networkStatus.ts` for that exact reason.

benjamn commented on src/core/QueryInfo.ts:
```diff
@@ -212,74 +165,100 @@ export class QueryInfo {
   }
 
   // This method can be overridden for a given instance.
   public cancel() {}
```
I'll investigate. By way of explanation, the `QueryInfo` class has become much less central/important with this PR, but I haven't fully examined what's left to make sure it's all necessary and tidy.

hwillson replied:
For sure - totally understood. This definitely isn't pressing (or needed for AC3, or potentially even needed at all).

benjamn replied:
Resolved by 6122cac3cb5b6e7ec60ac745dc4ca16d89eeb463.

hwillson replied:
Oh nice, and you addressed the TODO - awesome!

benjamn commented on src/utilities/observables/observables.ts:
```diff
@@ -1,66 +1,54 @@
-      }
-    };
-  });
+export function multicast<T>(...sources: Observable<T>[]) {
```
Will do.

benjamn commented on src/core/Reobserver.ts:
```diff
@@ -0,0 +1,142 @@
+import { Concast } from '../utilities/observables/Concast';
+import { invariant } from 'ts-invariant';
+
+export class Reobserver<TData, TVars> {
```
Done! 190ce59b0790243e679baadb7c60a3b695ef163f

hwillson replied:
Awesome - looks great @benjamn, thanks!

benjamn commented on src/core/QueryManager.ts:
```diff
@@ -343,150 +347,16 @@ export class QueryManager<TStore> {
-      needToFetch && fetchPolicy !== 'cache-only' && fetchPolicy !== 'standby';
-
-    // we need to check to see if this is an operation that uses the @live directive
-    if (hasDirectives(['live'], query)) shouldFetch = true;
```
4e48779d0acd724488828549347e432852263413

benjamn commented:
@hwillson Ok, I believe I've addressed all your feedback. Does your ✅ of approval mean you're comfortable with, like, actually _merging_ this PR? Because I'm ready if you are…

hwillson commented:
@benjamn absolutely! 🎉
